### PR TITLE
FLINK-3398: Allow for opting-out from Kafka offset auto-commit

### PR DESCRIPTION
**Pull request by shikhar**

### CODE REVIEW

**Comment by rmetzger** [comment]
The change is breaking some tests.

**Comment by shikhar** [comment]
@rmetzger the test failures were because we had `auto.commit.enable=false` in the standard `KafkaTestEnvironmentImpl` standard props and it wasn't respecting that previously, updated those defaults

**Comment by shikhar** [comment]
Just fixed the `Kafka09ITCase.testCheckpointing()` fail as well, hopefully CI is green with that.

**Comment by rmetzger** [comment]
Thank you for working on this. I've been very busy with the 1.0 release, vacation and some talks, but now I'm back for my regular work.
I try to review the PR today or on Monday.

File: flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer08.java [hunk 1]
```diff
@@ -295,7 +296,7 @@ public void open(Configuration parameters) throws Exception {
 
 	@Override
 	public void run(SourceContext<T> sourceContext) throws Exception {
-		if (fetcher != null) {
+		if (fetcher != null && isAutoCommitEnabled()) {
```
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer08.java]: This is not correct!
If the fetcher != null and auto commit disabled, `fetcher.run()` (line 316) is never called! This means that the fetcher is doing anything at all!

I'm not sure how the integration tests were passing with that change! (maybe there is no test testing auto commit enable = false.)

**Comment by rmetzger** [comment]
Can you add a test case ensuring the functionality? If I'm not mistaken, the change would have introduced a pretty bad bug.

File: flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer08.java (hunk 1 above)
Comment (shikhar) [inline flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer08.java]: Ikes! Added that on the wrong conditional, it was supposed to be on the `!streamingRuntimeContext.isCheckpointingEnabled()` clause where `PeriodicOffsetCommitter` is initialized.

**Comment by rmetzger** [comment]
Let me summarize the change:

```
auto.commit.enable=true checkpointing=true    offsets go to ZK/Broker on each completed checkpoint
auto.commit.enable=false checkpointing=true   offsets are only available for recovery with Flink, not in ZK / broker
auto.commit.enable=true checkpointing=false   offsets are written to ZK periodically
auto.commit.enable=false checkpointing=false  offsets are not in ZK or in Flink
```

I don't like this variant:
```
auto.commit.enable=false checkpointing=true   offsets are only available for recovery with Flink, not in ZK / broker
```
because it stretches the original definition of "auto.commit.enable" a bit. In my understanding, "auto.commit.enable" in Kafka starts a periodic committer, for sending the offsets to ZK / broker.

In Flink, the semantic of the setting is more or less "enable offset committing".
How about we limit the "auto.commit.enable" setting only to the periodic offset committing thread?

**Comment by StephanEwen** [comment]
I think the functionality is very desirable. In many cases, you really don't want Flink to write something to Zookeeper for some random GroupId. This floods ZooKeeper with garbage that helps no one.

I think that this is actually a kind of nice solution. The `auto.commit.enable` in Flink means not necessary periodically, but "on checkpoint or periodically".

**Comment by wanderingbort** [comment]
I just wanted to echo @StephanEwen 's sentiment.  

Flink's management of offsets seems to supersede the Kafka concept of committed offsets (whether to ZK or an 0.9 broker) leaving me to think there are only 2 scenarios where offset commitment is useful in the context of flink :

- coordinating with delivery with consumers outside the flink ecosystem (which seems conceptually dubious)
- reporting to monitoring software that understands 0.8- zookeeper stored offsets or 0.9+ broker stored offsets

I would consider both of these to be optional features and not required features. 

For my particular use case, we are on Kafka 0.8 and requiring write access to the zookeeper cluster that is coordinating kafka in order to access read-only message data is problematic.

**Comment by shikhar** [comment]
Another thing is that presence of offsets in ZK (0.8) / Kafka (0.9) also affects what happens when a job is starting without a checkpoint / savepoint. If they are present, the reset mode won't be respected, and there is potential for dropping messages (e.g. if the recovery strategy in the absence of a valid checkpoint or savepoint is to process from the beginning of available data with offset reset mode 'earliest').

**Comment by tzulitai** [comment]
What's the status of this PR? I think it is a good functionality to add. @shikhar if you'd like to continue working on this, I'm happy to help shepherd this PR for merging.

Some of my input from the previous discussions on the interpretation of "auto.commit.enable":
I agree with @rmetzger that stretching the definition of this config to "on checkpoint or periodically" isn't a good idea. Committing the offsets back to Kafka on 'notifyCheckpointComplete()' is a Flink-specific behaviour. I think mixing Flink-specific behaviour within Kafka original configs can be confusing for frequent Kafka users and difficult to maintain in the long run.

Perhaps we can have something like "flink.commit-on-checkpoint". What I have in mind:
> // commit to ZK / broker only on Flink's checkpoint.
> flink.commit-on-checkpoint=true auto.commit.enable=*
>
> // periodically commits to ZK / broker
> flink.commit-on-checkpoint=false auto.commit.enable=true
>
> // no commit to ZK / broker at all
> flink.commit-on-checkpoint=false auto.commit.enable=false

The first option will be valid only when checkpointing is enabled also. Otherwise, whether or not checkpointing is enabled does not effect the above settings.

I think this better matches the idea that "the Kafka offset store is only used for exposure of progress to the outside world, and not for recovery / manipulation of how Flink reads from Kafka".

**Comment by shikhar** [comment]
Thanks @tzulitai! I like your proposal, I'm afraid I won't be able to get around to shaping up this PR though so might be best to start a new one (please feel free to close).

**Comment by tzulitai** [comment]
@shikhar Sure, please feel free to tag me for a review once you are done. I think you'll need to manually close this PR yourself, you can close it when the new PR is opened. Another option is to rebase this PR on the current master, and after you finish shaping up again, force push to this same PR branch.

Also, since this will be affecting how users interact with the Kafka consumer, let's wait until we get consensus from @StephanEwen and @rmetzger on the proposed change before you get on to working on it ;)

**Comment by shikhar** [comment]
@tzulitai sorry I meant I won't be able to get to this change at all, so anyone please feel free to take this up!

**Comment by tzulitai** [comment]
@shikhar Ok, sure no problem :)

**Comment by StephanEwen** [comment]
I left a few thoughts in the corresponding JIRA issue: https://issues.apache.org/jira/browse/FLINK-3398

**Pull request closed without merging**